File: c++-library/fenwick_tree.cpp
/**
 * @brief ange Sum Query by FenwickTree(Binary Indexed Tree)
 *
 * total number: n
 * queries:
 *    1. add(i, val): bit[i] += val
 *    2. sum(n): sum(bit[0] + ... + bit[n-1])
 * query time complexity: O(log n)
 * space complexity: O(N)
 *
 * Self-balancing binary search tree or Segment Tree can do the same, it takes
 * longer to program and complexity also increases.

 * Thanks: http://hos.ac/slides/20140319_bit.pdf
 * used in ARC031 C, indeednow finalB E, DSL2B(AOJ), ARC033 C, yukicoder No.59
 */

#include <cassert>
#include <iostream>
#include <vector>

template <typename T>
struct RangeSumQuery {
  int N;
  std::vector<T> dat;

  RangeSumQuery(int N) : N(N) {
    assert(N > 0);
    dat.resize(N, 0);
  }

  void add(int k, T val) {
    assert(0 <= k && k < N);
    for (int x = k; x < N; x |= x + 1) {
      dat[x] += val;
    }
  }

  // [0, k)
  T sum(int k) const {
    assert(k >= 0);
    T ret = 0;
    for (int x = k - 1; x >= 0; x = (x & (x + 1)) - 1) {
      ret += dat[x];
    }
    return ret;
  }
  // [l, r)
  T sum(int l, int r) const {
    assert(0 <= l && l <= r && r <= N);
    return sum(r) - sum(l);
  }

  T index(int k) const {
    assert(0 <= k && k < N);
    return sum(k + 1) - sum(k);
  }
};

void yosupo() {
  int N, Q;
  std::cin >> N >> Q;
  RangeSumQuery<long long> rsq(N);
  for (int i = 0; i < N; i++) {
    long long v;
    std::cin >> v;
    rsq.add(i, v);
  }

  for (int i = 0; i < Q; i++) {
    int t;
    long long x, y;
    std::cin >> t >> x >> y;
    if (t == 0) {
      rsq.add(x, y);
    } else {
      std::cout << rsq.sum(x, y) << '\n';
    }
  }
}

int main() {
  std::cin.tie(0);
  std::ios_base::sync_with_stdio(false);

  yosupo();

  return 0;
}

File: python_library_typed/data_structures/fenwick_tree.py
import math
from typing import List


class FenwickTree1:
    """
    FenwickTree(Binary Indexed Tree)
    total number: n
    queries:
    1. add(i, val): add val to i-th value
    2. sum(n): sum(bit[1] + ... + bit[n])
    complexity: O(log n)

    Self-balancing binary search tree or Segment Tree can do the same,
    it takes longer to program and complexity also increases.
    Thanks: http://hos.ac/slides/20140319_bit.pdf

    used in ARC031 C, indeednow finalB E
    """
    def __init__(self, a_list: List[int]) -> None:
        # 1-index
        self.N: int = len(a_list)
        self.bit: List[int] = a_list[:]
        for i in range(1, self.N):
            self.bit[i + (i & -i)] += self.bit[i]

    def add(self, idx: int, val: int) -> None:
        while idx <= self.N:
            self.bit[idx] += val
            idx += idx & -idx

    def sum(self, n: int) -> int:
        ret: int = 0
        while n > 0:
            ret += self.bit[n]
            n -= n & -n
        return ret


class FenwickTree:
    def __init__(self, a_list: List[int]) -> None:
        # 0-indexed
        self.N: int = len(a_list)
        self.bit: List[int] = a_list[:]
        for _ in range(self.N, 1 << (math.ceil(math.log(self.N, 2)))):
            self.bit.append(0)
        for i in range(self.N-1):
            self.bit[i | (i+1)] += self.bit[i]

    def add(self, i: int, val: int) -> None:
        while i < self.N:
            self.bit[i] += val
            i |= i + 1

    def sum(self, n: int) -> int:
        ret: int = 0
        while n >= 0:
            ret += self.bit[n]
            n = (n & (n + 1)) - 1
        return ret
